### PR TITLE
feat(server): Add inbound filters functionality to dynamic sampling

### PR DESCRIPTION
**Pull request by RaduW**
Adds dynamic sampling rules to support inbound filters functionality.

Covering the inbound filters functionality with dynamic sampling rules allows us to remove
inbound filters and present the user with only one concept and one configuration interface.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 ## Unreleased
 
+**Features**:
+- Add inbound filters functionality to dynamic sampling rules. ([#920](https://github.com/getsentry/relay/pull/920))
```
Comment (jan-auer) [inline CHANGELOG.md]: Let's move this to `Internal` again. Dynamic sampling is not an exposed feature yet. This aligns with how we communicated the past dynamic sampling PRs if we compare with the last releases.

File: relay-filter/src/browser_extensions.rs [hunk 2]
```diff
@@ -7,24 +7,32 @@ use relay_general::protocol::{Event, Exception};
...
 use crate::{FilterConfig, FilterStatKey};
 
-/// Filters events originating from known problematic browser extensions.
-pub fn should_filter(event: &Event, config: &FilterConfig) -> Result<(), FilterStatKey> {
-    if !config.is_enabled {
-        return Ok(());
-    }
-
+/// Check if the event originates from known problematic browser extensions.
+pub fn mathces(event: &Event) -> bool {
```
Comment (jan-auer) [inline relay-filter/src/browser_extensions.rs]:
```suggestion
pub fn matches(event: &Event) -> bool {
```

Comment (jan-auer) [inline relay-filter/src/browser_extensions.rs]: For all `matches` functions, please add the filter config to the signature so that all filters have the same signature for `matches` and `should_filter`. The only difference then is that one returns a result and the other returns a `bool`. Please change this in all filters.

On a side note, the names `should_filter` and `matches` don't necessarily make it clear what the difference is. Under the assumption that you're trying not to change `should_filter` in this PR, we can stick with the current naming.

File: relay-filter/src/client_ips.rs [hunk 3]
```diff
@@ -10,6 +10,27 @@ use ipnetwork::IpNetwork;
 
 use crate::{ClientIpsFilterConfig, FilterStatKey};
 
+/// Checks if an ip address is in a list of blacklisted ip addresses
```
Comment (jan-auer) [inline relay-filter/src/client_ips.rs]: Please either use the same doc string as in `should_filter`, or move the documentation to the module level and link to it from the function docs.

File: relay-filter/src/config.rs [hunk 4]
```diff
@@ -47,14 +47,9 @@ pub enum LegacyBrowser {
...
 
-impl<'de> Deserialize<'de> for LegacyBrowser {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: serde::de::Deserializer<'de>,
-    {
-        let string = Cow::<str>::deserialize(deserializer)?;
-
-        Ok(match string.as_ref() {
+impl From<&str> for LegacyBrowser {
```
Comment (jan-auer) [inline relay-filter/src/config.rs]: Please implement `FromStr`, instead.

File: relay-filter/src/csp.rs [hunk 5]
```diff
@@ -38,8 +49,11 @@ pub fn should_filter(event: &Event, config: &CspFilterConfig) -> Result<(), Filt
 /// or None (matches anything in the respective position)
 #[derive(Hash, PartialEq, Eq)]
 pub struct SchemeDomainPort {
+    /// the scheme of the url
```
Comment (jan-auer) [inline relay-filter/src/csp.rs]: Same for the following doc comments.

```suggestion
    /// The scheme of the url.
```

File: relay-general/src/protocol/session.rs [hunk 6]
```diff
@@ -217,7 +217,7 @@ mod tests {
         let mut parsed = SessionUpdate::parse(json.as_bytes()).unwrap();
 
         // Sequence is defaulted to the current timestamp. Override for snapshot.
-        assert!((parsed.sequence - default_sequence()) <= 1);
+        assert!((default_sequence() - parsed.sequence) <= 1);
```
Comment (jan-auer) [inline relay-general/src/protocol/session.rs]: Good catch! Since this is an unrelated change, can you please create a separate PR for this and add a description on the change?

**Review by jan-auer — COMMENTED** [review]
Partial review of just the filter changes.

File: relay-filter/src/config.rs (hunk 4 above)
Comment (RaduW) [reply relay-filter/src/config.rs]: Are we sure? 
I thought that `FromStr` is used for parsing that can fail. 
Since we convert any string into a `LegacyBrowser` ( either a known one or a `Unknown(String)` ) having to deal with a `Result<LegacyBrowser, Self::Err>` seems quite un-ergonomic for no apparent benefit.

File: CHANGELOG.md (hunk 1 above)
Comment (RaduW) [reply CHANGELOG.md]: ok

File: tests/integration/fixtures/processing.py [hunk 7]
```diff
@@ -41,6 +41,7 @@ def inner(options=None):
         if processing.get("kafka_config") is None:
             processing["kafka_config"] = [
                 {"name": "bootstrap.servers", "value": bootstrap_servers},
+                {"name": "broker.address.family", "value": "v4"},
```
[the remark this answers is not in the document]
Comment (RaduW) [reply tests/integration/fixtures/processing.py]: already done (maybe you are looking at a stale version )?

File: relay-filter/src/browser_extensions.rs (hunk 2 above)
Comment (RaduW) [reply relay-filter/src/browser_extensions.rs]: thanks

File: relay-filter/src/client_ips.rs (hunk 3 above)
Comment (RaduW) [reply relay-filter/src/client_ips.rs]: ok, reviewed all `matches` comments, and changed to contain same info as `should_filter`

File: relay-filter/src/csp.rs (hunk 5 above)
Comment (RaduW) [reply relay-filter/src/csp.rs]: yep

File: relay-general/src/protocol/session.rs (hunk 6 above)
Comment (RaduW) [reply relay-general/src/protocol/session.rs]: ok

File: relay-filter/src/config.rs (hunk 4 above)
Comment (jan-auer) [reply relay-filter/src/config.rs]: Yeah, this is an interesting case since we express failure to parse by returning `Unknown`. The standard library uses `Infallible` as error in such cases, for example in [`impl FromStr for String`](https://doc.rust-lang.org/stable/std/str/trait.FromStr.html#impl-FromStr-36).

Otherwise, I'd prefer an explicit infallible `parse` method instead of the conversion trait.

File: tests/integration/fixtures/processing.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jan-auer) [reply tests/integration/fixtures/processing.py]: Sorry, that was from an old review that GH kept in local storage.

File: relay-filter/src/client_ips.rs (hunk 3 above)
Comment (jan-auer) [reply relay-filter/src/client_ips.rs]: Can you please check again? There's still a difference `browser_extensions`, at least.

Comment (jan-auer) [reply relay-filter/src/client_ips.rs]: Never mind, you use different language for "check" and "filter". WFM 👍

File: relay-server/src/utils/dynamic_sampling.rs [hunk 8]
```diff
@@ -7,9 +7,15 @@ use futures::{future, prelude::*};
 use rand::{distributions::Uniform, Rng};
 use rand_pcg::Pcg32;
 use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::collections::HashMap;
+use std::net::IpAddr;
```
Comment (jan-auer) [inline relay-server/src/utils/dynamic_sampling.rs]: _nit_: Please sort imports according to our style guide.

File: relay-server/src/utils/dynamic_sampling.rs [hunk 9]
```diff
@@ -127,8 +201,11 @@ impl NotCondition {
     fn supported(&self) -> bool {
         self.inner.supported()
     }
-    fn matches<T: FieldValueProvider>(&self, value_provider: &T) -> bool {
-        !self.inner.matches(value_provider)
+    fn matches_event(&self, event: &Event, ip_addr: Option<IpAddr>) -> bool {
```
Comment (jan-auer) [inline relay-server/src/utils/dynamic_sampling.rs:204]: In a follow-up, let's replace all these `ip_addr` with RequestMeta: 
- It's to be expected that we require more information about the request
- `ip_addr` doesn't specify which IP we want: the socket, the client, how to handle x-forwarded-for? This is what RequestMeta already solves for us.

File: relay-server/src/utils/dynamic_sampling.rs [hunk 10]
```diff
@@ -86,6 +128,24 @@ impl GlobCondition {
     }
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CustomCondition {
+    pub name: String,
+    #[serde(default)]
+    pub value: Value,
```
Comment (jan-auer) [inline relay-server/src/utils/dynamic_sampling.rs:135]: If we have generic options, why do we need `value` here?

For a follow-up: It would be interesting to type out the custom conditions, since they are statically known in the code.

File: relay-server/src/utils/dynamic_sampling.rs [hunk 11]
```diff
@@ -192,24 +283,37 @@ impl SamplingRule {
...
-    fn get_value(&self, path: &str) -> FieldValue;
+    fn get_value(&self, path: &str) -> Value;
     /// what type of rule can be applied to this provider
     fn get_rule_type(&self) -> RuleType;
+    /// returns a filtering function for custom operators.
+    /// The function returned takes the provider and a condition definition and
+    /// returns a match result
+    fn get_custom_operator(
+        name: &str,
+    ) -> fn(condition: &CustomCondition, slf: &Self, ip_addr: Option<IpAddr>) -> bool;
```
Comment (jan-auer) [inline relay-server/src/utils/dynamic_sampling.rs:294]: Probably also for a follow-up: Can we make it so that the Field provider inherently has access to the `ip_addr` (or, better, the entire RequestMeta)? We could implement the trait on a type like this:

```
struct EventValueProvider<'a> {
  event: &'a Annotated<Event>,
  meta: &'a RequestMeta,
}
```

Comment (jan-auer) [inline relay-server/src/utils/dynamic_sampling.rs:294]: Q: Why not just run the custom condition inside the function? 

```rust
fn match_custom(&self, condition: &CustomCondition, ip_addr: Option<IpAddr>) -> bool;
```

For a follow-up: Later down the road we need to refactor the value provider so that it can expose all fields needed by the custom rules via its `get_value` interface.

File: relay-server/src/utils/dynamic_sampling.rs (hunk 10 above)
Comment (RaduW) [reply relay-server/src/utils/dynamic_sampling.rs:135]: to have a consistent structure that  matches with the non custom operators.
For example:
```
{ "name": "event.releases", "op": "glob", "value": ["v1.1", "v2.*"]},
{ "name": "event.legacy_browsers", "op":"custom", "value":["ie8", "ie9"]},

```
The options are supposed to be used for configuring the operator (something like the `ignore_case` from operator `eq` ).

Comment (jan-auer) [reply relay-server/src/utils/dynamic_sampling.rs:135]: Understood. Once these are typed out it certainly looks better again. Let's keep like this, then.

Comment (jan-auer) [reply relay-server/src/utils/dynamic_sampling.rs:135]: On another thought: It seems odd that we have `options` but `name` and `value` special cased on the top, once everything is typed out. We can address this when typing out, though.

File: relay-server/src/utils/dynamic_sampling.rs (hunk 11 above)
Comment (RaduW) [reply relay-server/src/utils/dynamic_sampling.rs:294]: We could do that, it seemed wrong though. 
The ValueProvider gives you information from the Event or Trace but doesn't also run 
rules, anyway not feeling too strongly either way.
The presence of the custom operation in ValueProvider makes it quite a messy abstraction.

File: relay-filter/src/config.rs (hunk 4 above)
Comment (RaduW) [reply relay-filter/src/config.rs]: ok, done

File: relay-server/src/utils/dynamic_sampling.rs (hunk 11 above)
Comment (jan-auer) [reply relay-server/src/utils/dynamic_sampling.rs:294]: Separate from this PR, could you compile a list of all attributes that we need access to for running the custom matchers?

File: relay-server/src/utils/dynamic_sampling.rs [hunk 12]
```diff
@@ -220,29 +325,111 @@ impl FieldValueProvider for Event {
...
+}
+
+fn legacy_browsers_matcher(
+    condition: &CustomCondition,
+    event: &Event,
+    _ip_addr: Option<IpAddr>,
+) -> bool {
+    let browsers = condition.value.as_array().map(|v| {
+        v.iter()
+            .map(|s| LegacyBrowser::from_str(s.as_str().unwrap_or("")).unwrap())
```
Comment (jan-auer) [inline relay-server/src/utils/dynamic_sampling.rs]: _nit_: Thanks to `FromStr`, you can now use `parse`:

```suggestion
            .map(|s| s.as_str().unwrap_or("").parse().unwrap())
```

File: relay-server/src/utils/dynamic_sampling.rs [hunk 13]
```diff
@@ -586,10 +807,122 @@ mod tests {
...
         for (rule_test_name, condition) in conditions.iter() {
             let failure_name = format!("Failed on test: '{}'!!!", rule_test_name);
-            assert!(condition.matches(&tc), failure_name);
+            assert!(condition.matches_trace(&tc, None), failure_name);
         }
     }
 
+    #[test]
+    /// test matching for various rules
+    fn test_matches_events() {
```
Comment (jan-auer) [inline relay-server/src/utils/dynamic_sampling.rs:815]: In a follow-up, I would like to move all these tests to macro-based expansion. That way, each test runs in its own test function and we can see individual failures more easily.

We could also start adopting a test parameterization framework, but I'm still not convinced of that.

**Review by jan-auer — APPROVED** [review]

File: relay-server/src/utils/dynamic_sampling.rs (hunk 13 above)
Comment (RaduW) [reply relay-server/src/utils/dynamic_sampling.rs:815]: While I like the idea in principle it would make debugging difficult/impossible and I like to have this option open. 
I've been looking, a while back now, for a clean way to do parameterization of unit tests in Rust but all I found were frameworks that either used nightly builds or used some ugly macros.

If we can find a relatively clean way to do it while still being able to run a debugger I all for it.

Comment (jan-auer) [reply relay-server/src/utils/dynamic_sampling.rs:815]: You can run and debug tests defined in macros just like any other test. Rust Analyzer even provides code lenses for debugging individual tests, for example as we do in [`relay-general/tests/test_fixtures.rs`](https://github.com/getsentry/relay/blob/569678659bac656d6c4a6edcbdf3381677b5eaa2/relay-general/tests/test_fixtures.rs#L102-L107):

image
